src/resizing.jl

```diff
@@ -317,6 +334,12 @@ function imresize!(resized::AbstractArray{T,N}, original::AbstractArray{S,N}) wh
     imresize!(resized, itp)
 end
```

Comment:
Instead of adding more and more specifications, we could instead generalize our `imresize` API (actually you're already doing it) to pass extra args to `imresize!`, and let `imresize!` decide how to deal with it.

Unfortunately, we can't make `method` a positional argument because there's a hard-to-solve method ambiguity on `imresize(img, sizes...)`. Hence what I'm proposing is to introduce a keyword `method`:

- `imresize(img, args...; method=BSpline(Linear()))`
- `imresize!(img, args...; method=BSpline(Linear()))`.

```diff
- function imresize(original::AbstractArray{T,N}, new_size::Dims{N}) where {T,N}
+ function imresize(original::AbstractArray{T,N}, new_size::Dims{N}; kwargs...) where {T,N}
    Tnew = imresize_type(first(original))
    inds = axes(original)
    if map(length, inds) == new_size
        dest = similar(original, Tnew, new_size)
        if axes(dest) == inds
            copyto!(dest, original)
        else
            copyto!(dest, CartesianIndices(axes(dest)), original, CartesianIndices(inds))
        end
    else
-        imresize!(similar(original, Tnew, new_size), original)
+        imresize!(similar(original, Tnew, new_size), original; kwargs...)
    end
end

- function imresize!(resized::AbstractArray{T,N}, original::AbstractArray{S,N}) where {T,S,N}
+ function imresize!(resized::AbstractArray{T,N}, original::AbstractArray{S,N}; method::InterpolationType = BSpline(Linear())) where {T,S,N}
    # FIXME: avoid allocation for interpolation
-    itp = interpolate(original, BSpline(Linear()))
+   itp = interpolate(original, method)
    imresize!(resized, itp)
end
    # FIXME: avoid allocation for interpolation
    imresize!(resized, interpolate(original, method))
end
```

This way you don't need to copy-paste the existing methods. I didn't verify if this works out of the box, may I count on you to do this for other `imresize` and `imresize!` methods?

Reply:
We could also introduce `method` to `imrotate`, too, and that would just solve #111 if everything works.

Reply:
The main reason I did not do this was to maintain compatibility with the current form where for imrotate for example, you can pass Linear(), Cubic(OnGrid()), etc to degree. This becomes superfluous in the presence of method, but I suspect we should keep it for back compatibility. Thoughts on this?

Reply:
The main reason we made this a positional argument is because Julia was not performant dealing with keyword arguments. Now that since keyword argument overhead is optimized away, we could switch to keyword argument.

As you may have noticed, positional arguments make method dispatching quite complicated to reason about and extend.

Whether we want to keep backward compatibility totally depends on you; we could bump a minor version if you don't want to spend that much time dealing with it. If you choose to keep backward compatibility, which is great, you then need to add some if-branches to check values, and manually throw deprecation warnings. An example of such can be found in [Documenter](https://github.com/JuliaDocs/Documenter.jl/blob/b5356729147f269d8fc8cb3412f7630b8edd3bf3/src/Documenter.jl#L358-L377)

Reply:
An intermediate option I want to propose is just to keep the ``degree`` versions of each function and add ``kwargs...`` so that the ``method`` kwarg is overridden by the existence of a degree keyword. The only problem with that for imrotate is the function on line 27 of imresizing.jl. Can you explain better why that case cannot be taken care of by just setting the default behavior for when only an array is passed to be a simple linear interpolation? Why is there so much type handling going on here?

Reply:
> The only problem with that for imrotate is the function on line 27 of imresizing.jl. Can you explain better why that case cannot be taken care of by just setting the default behavior for when only an array is passed to be a simple linear interpolation? Why is there so much type handling going on here?

There's no `imresizing.jl` here... Hmmm, which method are you asking?

Reply:
> An intermediate option I want to propose is just to keep the degree versions of each function and add kwargs... so that the method kwarg is overridden by the existence of a degree argument.

Could you give an example of your proposed API? Is it this?

```julia
imresize(img, sizes...; degree = Linear(), method = BSpline(Linear()))
```

It will be a little bit mind puzzling which keyword argument has higher priority, so what about:

```julia
imresize(img, sizes...; method = Linear())
```

with `method` accept both `InterpolationType` and `Degree`? When passing a degree, it is assumed to be BSpine.

Reply:
> > The only problem with that for imrotate is the function on line 27 of imresizing.jl. Can you explain better why that case cannot be taken care of by just setting the default behavior for when only an array is passed to be a simple linear interpolation? Why is there so much type handling going on here?
> 
> There's no `imresizing.jl` here... Hmmm, which method are you asking?

My bad, I meant interpolations.jl

Reply:
My proposed API is the following for the box_extrapolation which underlies the imrotate. Something similar for resize.

``box_extrapolation(parent::AbstractArray, args...; method::Interpolations.InterpolationType=BSpline(Linear()), kwargs...)``

but if you pass a degree argument you get the original function definition which ignores the kwarg.

``box_extrapolation(parent::AbstractArray{T,N}, degree::Interpolations.Degree, args...; kwargs...)``

The only part I am not sure this handles gracefully is the following

``box_extrapolation(parent::AbstractArray, degree::D = Linear(), args...; kwargs...) where D<:Union{Linear,Constant}``

Reply:
It seems this does a nice job of preserving ``===`` in the tests for linear and constant interpolations which are broken in my version without a replacement for that last function...

Reply:
I have no idea, either. My best guess is that it deals with cases such as `imresize(img, fill = 0)` where `img` is a colorant array.

This part of the code was originally written by @Evizero even before Julia 0.7 so there're a lot of decisions that get outdated and unnecessary now. I would suggest rewriting `box_extrapolation` for cleaner logic and take advantage of the keyword arguments. @Evizero has written a very verbose set of test cases so rewriting it should be pretty safe.

My previous argument on `imresize` should also be valid for `box_extrapolation`:

```julia
# method can either be `Degree` or `InterpolationType`; degree is automatically wrapped with `BSpine`.
box_extrapolation(parent::AbstractArray{T,N}; method= Linear(), fill = _default_fill(T)) where {T, N}
```

My guess is that we don't need `box_extrapolation` for either `AbstractInterpolation` and `AbstractExtrapolation`. These two methods are implementation details and can thus be removed; if you search `box_extrapolation` in other files, you'll see all calls to `box_extrapolation` are `AbstractArray`s.

Reply:
I am hazy on the details, but I know I used to be quite fond of language-version specific workarounds to get the most out of the code. I would not be surprised if whatever motivated my designed choices has long been resolved on a language level.

So from my perspective feel free to question my code and rewrite it in a way that it makes sense today.
Think outside the `box_` :-)